### PR TITLE
Expose history function in placement storage interface

### PR DESCRIPTION
**Pull request by cw9**
So we can read placement in the history
@prateek @xichen2020

### CODE REVIEW

**Comment by xichen2020** [comment]
Hm why do we need to export this? I'm actually in favor of removing the `History` API altogether since it's semantics is not well defined and in general seems like an overly strong requirement for a general key value store

**Comment by prateek** [comment]
Chao put this up because of something I ran into this afternoon. A join node in M3DB triggered a panic. I was trying to look at past placements to see shard distribution for a theory and we figured it'd be useful to have an API to enable this in M3Cluster directly. 

Similar to this case, I'm sure people would like the ability to look at a bounded history of changes per key. Lack of it would make debugging/auditing changes made to KV much harder, if not impossible. 

Would defining explicit semantics (e.g. bounding to X historical entries per key) help allay the concern?

File: placement/types.go [hunk 1]
```diff
@@ -431,6 +431,9 @@ type Storage interface {
 
 	// Proto returns the placement proto.
 	Proto() (proto.Message, int, error)
+
+	// History returns the placement of a specific version.
+	History(version int) (Placement, error)
```
Comment (prateek) [inline placement/types.go]: Do we need a way to get the past version numbers themselves? Or are versions always incremented?

**Comment by xichen2020** [comment]
Thanks for the usecase, that helps to provide some context for the change.

Just to elaborate a little, I have two main concerns:
* The `History` API as part of the `Store` interface puts too strong a requirement on the underlying implementation. It requires that all implementations of this interface keep track of multiple versions (otherwise the current API should do) or return a NotImplemented error, which doesn't feel very clean to me. For example, a Cassandra based implementation that only keeps the last version using `CheckAndSet` would need to return an error here. If there is really such a need, I'd suggest moving the `History` and its relatives to a separate `MultiVersionedStore` interface and pull that out of the generic `Store` interface. This way when we move the KV related interface and implementation out of m3cluster into a standalone repo just for KV stuff, it'll make a lot more sense and requires less refactoring.

* The semantics for `History` is not well defined. Specifically, each implementation of the KV store may have their own way of retaining versions, if at all. For example, an implementation that only keeps the last version can only return a valid result for the last version. Another implementation that compacts data aggressively (e.g., etcd) may return different results depending on the compaction settings and when the `History` API is called. Yet another implementation that retains all histories indefinitely will always return valid results. This means different implementations of the KV store will have different semantics for the `History` API, which IMO puts a lot of cognitive burden on the users of this API since they essentially need to know the peculiarities of the different implementations since the semantics is inconsistent across implementations.

File: placement/types.go (hunk 1 above)
Comment (cw9) [reply placement/types.go]: versions on the placement are always incremented, when you get the latest placement you will see the current version, then you can refer the previous versions you want to look up

**Comment by cw9** [comment]
I agree that the history API could be pulled into MultiVersionedStore for better isolation, but I don't understand the concern about different implementations, if the store contains the version(s) requested, it returns the value(s) for the version(s), if the store does not contain the version(s) requested, for example the versions(s) could be even be invalid, or the history has been compacted during specific implementation, then the store should return errors. As long as we define a clear semantic on the interface then different implementations would just follow the semantic, no?

**Comment by xichen2020** [comment]
You could define the semantic of the `History` API that way yes, that will be consistent across implementations, but IMO the semantic is a bit unconventional. That said, I'd be less concerned if it's pulled into its own interface.

**Comment by cw9** [comment]
Sure, the original History API was mainly crafted for R2 usage, we could revisit the semantic there , I feel changing the interface from `History(from, to)` to `GetVersion(version)` is probably cleaner, but that should be a seperate diff. 

That said, I'm also going to rename the new function in placement storage from `History()` to `PlacementForVersion` to decouple the function names

File: placement/storage/helper.go [hunk 2]
```diff
@@ -74,6 +77,20 @@ func newPlacementHelper(store kv.Store, key string) Helper {
...
 }
 
+func (h *placementHelper) PlacementForVersion(version int) (placement.Placement, error) {
+	values, err := h.store.History(h.key, version, version+1)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(values) != 1 {
+		return nil, fmt.Errorf("invalid number of values returned: %d, expecting 1", len(values))
```
Comment (xichen2020) [inline placement/storage/helper.go]: `invalid number of placements` might be more clear.

File: placement/storage/helper.go [hunk 3]
```diff
@@ -74,6 +77,20 @@ func newPlacementHelper(store kv.Store, key string) Helper {
...
+	}
+
+	p, err := placementFromValue(values[0])
```
Comment (xichen2020) [inline placement/storage/helper.go]: nit: `return placementFromValue(values[0])`

**Pull request merged**